Allow not implemented tests in ExUnit

hubertlepicki opened the pull request:
My usual workflow in, say Rails, is that I first define a bunch of stub tests that look like this:

    test "some behavior"
    test "some other behavior"
    test "something else, maybe"

I have tried that with ExUnit and it exploded with undefined method test/1

This code creates a macro, which handles this situation. Such defined tests will be skipped by the test runner, and appropriate "Not yet implemented" message should be shown.

Please note it is like my first Elixir code I have written, if this is stupid it's totally my fault.

josevalim commented:
@hubertlepicki skipped tests are not shown in the report. In fact, the build will be considered successful. I don't believe this to be what we want?

hubertlepicki commented:
@josevalim I am not sure if that's correct apprach. There seems not to be "pending" state for tests as far as I can tell.

What I usually get is:

    66 tests, 0 failures, 3 pending

In this implementation you would see:
    66 tests, 0 failures, 3 skipped

So maybe what you say makes sense and pending should be different state than skipped.

In any case, I'd expect my build to succeed, I think

josevalim commented:
@hubertlepicki and, in those cases, are the test names that are pending printed?

hubertlepicki commented:
@josevalim no, they are not:

    $ mix test                
    ....

    Finished in 0.05 seconds (0.05s on load, 0.00s on tests)
    7 tests, 0 failures, 3 skipped

Do you think this is correct behavior, or should I dig deeper to add the listing of the skipped/not yet implemented tests?

josevalim commented:
Sorry, I meant to ask if they are shown in your Rails workflow. :)

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

hubertlepicki commented:
Actually I lied to you @josevalim , in Rails' default testing framework, not implemented tests raise failure as here:

    .......F
    
    Finished in 0.247608s, 32.3091 runs/s, 56.5410 assertions/s.
    
      1) Failure:
    PostTest#test_future_behavior_here [/home/hubert/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.2/lib/active_support/testing/declarative.rb:19]:
    No implementation provided for future behavior here
    
    8 runs, 14 assertions, 1 failures, 0 errors, 0 skips

I do use too much RSpec :(. There, the behavior is different:

    $ rspec spec
    
    Post
      should have some future behavior (PENDING: Not yet implemented)
    
    Pending: (Failures listed here are expected and do not affect your suite's status)
    
      1) Post should have some future behavior
         # Not yet implemented
         # ./spec/models/post_spec.rb:4
    
    
    Finished in 0.00048 seconds (files took 1.28 seconds to load)
    1 example, 0 failures, 1 pending

What should it be in Elixir? :)

josevalim commented:
@hubertlepicki since you are, can you also check what happens to the tests if a setup_all fails (in both)? :P Thanks!

hubertlepicki commented:
That will be 5zł. But ok.

Rspec (before(:all):

    $ rspec spec

    Post
      should have some future behavior (FAILED - 1)

    Failures:

      1) Post should have some future behavior
         Failure/Error: raise ActiveRecord::RecordNotFound
         ActiveRecord::RecordNotFound:
           ActiveRecord::RecordNotFound
         # ./spec/models/post_spec.rb:5:in `block (2 levels) in <top (required)>'

    Finished in 0.00065 seconds (files took 1.23 seconds to load)
    1 example, 1 failure

    Failed examples:

    rspec ./spec/models/post_spec.rb:8 # Post should have some future behavior

and Test::Unit/minitest (setup_all):

    $ ruby -Itest test/models/post_test.rb
    Run options: --seed 42309

    # Running:

    F

    Finished in 0.012336s, 81.0659 runs/s, 81.0659 assertions/s.

      1) Failure:
    PostTest#test_future_behavior_here [/home/hubert/.rvm/gems/ruby-2.2.2/gems/activesupport-4.2.2/lib/active_support/testing/declarative.rb:19]:
    No implementation provided for future behavior here

    1 runs, 1 assertions, 1 failures, 0 errors, 0 skips

josevalim commented:
Ok, thank you! I will think a bit more about this and write a proposal tomorrow. :)

hubertlepicki commented:
Aye, capitan!

josevalim commented:
Ok, so in ExUnit, tests have four states:

* success - those are printed as a dot, if all pass, status is 0
* failure - those are printed as failures, if one fails, status is 1
* invalid - those are printed as ?, it does not affect the build status. Currently invalid is only used when setup_all fails to mark the tests we were not able to run. Since it is used only in this situation, we could change its name
* skipped - those are never printed, they also do not affect build status

I don't want to add a fifth state, because that would be way too many. We also don't want those to be considered success nor skipped. So we have three options:

1. Print them as failures. This is what Rails does.

2. Print them as invalid, after all, a test without an implementation is not valid. We would need to consider if they should be printed too, similar to what RSpec does, or just leave it hidden (as in the PR).

3. Rename invalid to pending. That would be ok because we could argue that, when setup_all fails, the tests are marked as pending because we couldn't run the setup_all. We still need to choose if we want and how to print pending tests or just leave them hidden.

For all purposes, 1. is the simplest.

hubertlepicki commented:
I would argue that if setup_all fails, I'd like to see the error, or at least my build on the CI should fail (i.e. return 1 from the Unix process).

Let me see what we can do, I think 1. might be both the simplest and the best after all...

josevalim commented:
> I would argue that if setup_all fails, I'd like to see the error, or at least my build on the CI should fail (i.e. return 1 from the Unix process).

That's what happens today. We add a failure for the setup_all, but all other tests are marked as invalid (and not shown because the failure would be the same for all of them).

hubertlepicki commented:
Alright, I have implemented that and squashed the commits nicely. Current output is:


    $ mix test
      1) test first behavior (EtogglTest)
         test/etoggl_test.exs:4
         ** (RuntimeError) Not yet implemented
         stacktrace:
           test/etoggl_test.exs:4
    
    
    
    Finished in 0.02 seconds (0.02s on load, 0.00s on tests)
    1 test, 1 failure
    
    Randomized with seed 258695


I am not sure about I should raise RuntimeError or define special error type for this behavior. I can see both solutions in ExUnit source code, so your advise here @josevalim would be helpful...

josevalim commented:
Maybe `flunk("Not yet implemented")` would be better? You could also do:

  reraise(ExUnit.AssertionError, [message: "Not yet implemented"], [])

The third argument sets an empty stacktrace as it is probably useless in this case.

hubertlepicki commented:
flunk() looks nicer indeed on the output:

      1) test first behavior (EtogglTest)
         test/etoggl_test.exs:4
         Not yet implemented

updated the PR already. Now need to read what flunk() does ;)

hubertlepicki commented:
Yeah I think it makes sense, it is failed assertion of sort, that the test should be implemented. Merge & champagne?

josevalim commented on lib/ex_unit/test/ex_unit_test.exs:
```diff
@@ -16,11 +16,13 @@ defmodule ExUnitTest do
         assert false
       end
+
+      test "not implemented test"
```
Can we define this in a separate place and also check the output printed? Thank you.

hubertlepicki replied:
this is now resolved (?)

hubertlepicki commented:
@josevalim extracted the tests to separated example and verify the message in captured output now.

josevalim commented on lib/ex_unit/lib/ex_unit/case.ex:
```diff
@@ -227,6 +227,26 @@ defmodule ExUnit.Case do
+
+  Provides a convenient macro that allows a test to be
+  defined with a string, but not yet implemented. The
+  resulting test will be skipped by the tests runner.
```
Docs are out of date.

hubertlepicki replied:
This has been now updated.

josevalim commented on lib/ex_unit/test/ex_unit_test.exs:
```diff
@@ -138,4 +138,21 @@ defmodule ExUnitTest do
   end
+
+  test "it produces error on not implemented tests" do
+    capture_io :stderr, fn ->
```
Do we need to capture stderr?

hubertlepicki replied:
you are right, this is not needed. I have copied over test example from the same file, where, btw, it is not needed too. I can make second PR to clean this up.

josevalim commented:
/cc @ericmj @alco @fishcakez thoughts on this feature?

lexmag commented:
:+1: great that it'll be failure and not anything else. I've seen few times people tend not to complete pending tests for a long period. :smile:

ericmj commented:
Sounds good

fishcakez commented:
I would use this feature. Could these bodyless tests get a tag? If it had a tag I would probably use it to both exclude not implement tests when working on a new case and to include all when hunting down not implemented tests.

josevalim commented:
@fishcakez brilliant idea. I will give you dibs on figuring out what the tag should be named. :P

josevalim commented:
@hubertlepicki perfect, let's do it in another PR then. :)

hubertlepicki commented:
@josevalim sure, as soon as I know what the tag should be named I can create another PR to add it. @fishcakez "pending" maybe?

fishcakez commented:
@hubertlepicki, I think @josevalim is trolling me because naming is hard and I purposely left the naming ambiguous :P. Maybe the `flunk/1` message and the tag should be more closely related so some wording is shared or the meaning is at least the same so its easier to remember.

hubertlepicki commented:
Then the tag could be "not_yet_implemented" I suppose... @josevalim ?

fishcakez commented:
:not_implemented is good I think

On Tuesday, 23 June 2015, Hubert Łępicki <notifications@github.com> wrote:

> Then the tag could be "not_yet_implemented" I suppose... @josevalim
> <https://github.com/josevalim> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/3419#issuecomment-114467605>.
>

alco commented:
Together with a custom tag this idea starts to make sense. Without the tag it wouldn't be much different from writing

```elixir
test "what to do here", do: raise("not implemented")
```

But with the tag I can see a workflow where I would have tests with `not_implemented` skipped by default during development but always enabled on the CI.

hubertlepicki commented:
Alright. Let me add this tag and update this PR.

hubertlepicki commented:
I have added :not_implemented tag. But, honestly, I do not know how to test if this tag is really added in the unit tests. The existing tests for tagging functionality rely... on providing block to the test itself and doing assertions inside. In this case the whole point is not providing the block.

Anyway, I am leaving this in separate commit https://github.com/elixir-lang/elixir/commit/e63d1eda78c84e0dea1e7d7c2ad289b0fc5553da

what do you guys think?

hubertlepicki commented:
This does work when running the filters too:

    Excluding tags: [not_implemented: "true"]
    
    Finished in 0.02 seconds (0.02s on load, 0.00s on tests)
    1 test, 0 failures, 1 skipped

hubertlepicki commented:
whoohoo

josevalim commented:
I have pushed an assertion to check the tag is properly set. :)

hubertlepicki commented:
Have noticed it here: https://github.com/elixir-lang/elixir/commit/702cfbf8708d7474d57cb0c57c6996cc5e58a98e

Good to know you can run assertions in the setup block (although probably not to be overused).
